### ISSUE TITLE
Could not install package because of Nuget V2 deprecation

### ISSUE THREAD

**Issue by sajjadrad** [GlitchEnzo/NuGetForUnity#383]
## Description
I got api deprecation error while installing a package

- **NuGet Package:** System.Buffers [4.4.0, ) or greater
- **NuGetForUnity Version:** 3.0.1
- **Unity Version:** 2020.1.3f1
- **Operating System:** macOS 11.2

```Unable to retrieve package list from http://www.nuget.org/api/v2/FindPackagesById()?id='System.Buffers'&$orderby=Version asc
System.Net.WebException: The remote server returned an error: (400) Bad Request.
  at System.Net.HttpWebRequest+<GetResponseFromData>d__240.MoveNext () [0x00152] in <525dc68fbe6640f483d9939a51075a29>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Net.HttpWebRequest.GetResponse () [0x00013] in <525dc68fbe6640f483d9939a51075a29>:0 
  at NugetForUnity.NugetHelper.RequestUrl (System.String url, System.String userName, System.String password, System.Nullable 1[T] timeOut) [0x000b9] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.GetPackagesFromUrl (System.String url, System.String username, System.String password) [0x0006c] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.FindPackagesById (NugetForUnity.NugetPackageIdentifier package) [0x000ee] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
UnityEngine.Debug:LogErrorFormat(String, Object[])
NugetForUnity.NugetPackageSource:FindPackagesById(NugetPackageIdentifier)
NugetForUnity.NugetPackageSource:GetSpecificPackage(NugetPackageIdentifier)
NugetForUnity.NugetHelper:GetOnlinePackage(NugetPackageIdentifier)
NugetForUnity.NugetHelper:GetSpecificPackage(NugetPackageIdentifier)
NugetForUnity.NugetHelper:InstallIdentifier(NugetPackageIdentifier, Boolean)
NugetForUnity.NugetHelper:Install(NugetPackage, Boolean)
NugetForUnity.NugetHelper:InstallIdentifier(NugetPackageIdentifier, Boolean)
NugetForUnity.NugetHelper:Install(NugetPackage, Boolean)
NugetForUnity.Nuge```

**Comment by thuwharris** [comment]
I've been getting the same issue with this setup:

- NuGet Package: JsonSubTypes 1.8.0
- NuGetForUnity Version: 3.0.1
- Unity Version: 2019.4.17f1
- Operating System: Windows 10 (10.0.19042)

```
Unable to retrieve package list from http://www.nuget.org/api/v2/FindPackagesById()?id='Microsoft.CSharp'&$orderby=Version asc
System.Net.WebException: The remote server returned an error: (400) Bad Request.
  at System.Net.HttpWebRequest.EndGetResponse (System.IAsyncResult asyncResult) [0x00058] in <ef151b6abb5d474cb2c1cb8906a8b5a4>:0 
  at System.Net.HttpWebRequest.GetResponse () [0x0000e] in <ef151b6abb5d474cb2c1cb8906a8b5a4>:0 
  at NugetForUnity.NugetHelper.RequestUrl (System.String url, System.String userName, System.String password, System.Nullable`1[T] timeOut) [0x000b9] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.GetPackagesFromUrl (System.String url, System.String username, System.String password) [0x0006c] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.FindPackagesById (NugetForUnity.NugetPackageIdentifier package) [0x000ee] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
UnityEngine.Debug:LogErrorFormat(String, Object[])
NugetForUnity.NugetPackageSource:FindPackagesById(NugetPackageIdentifier)
NugetForUnity.NugetPackageSource:GetSpecificPackage(NugetPackageIdentifier)
NugetForUnity.NugetHelper:GetOnlinePackage(NugetPackageIdentifier)
NugetForUnity.NugetHelper:GetSpecificPackage(NugetPackageIdentifier)
NugetForUnity.NugetHelper:InstallIdentifier(NugetPackageIdentifier, Boolean)
NugetForUnity.NugetHelper:Install(NugetPackage, Boolean)
NugetForUnity.NugetHelper:InstallIdentifier(NugetPackageIdentifier, Boolean)
NugetForUnity.NugetHelper:Install(NugetPackage, Boolean)
NugetForUnity.NugetHelper:InstallIdentifier(NugetPackageIdentifier, Boolean)
NugetForUnity.NugetWindow:DrawPackage(NugetPackage, GUIStyle, GUIStyle)
NugetForUnity.NugetWindow:DrawPackages(List`1)
NugetForUnity.NugetWindow:DrawOnline()
NugetForUnity.NugetWindow:OnGUI()
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
```

**Comment by nobnak** [comment]
Similar issue here:

- Nuget Package : StackExchange.Redis 2.2.4
- NugetForUnity : 3.0.1
- Unity Version  : 2020.3.1f1
- OS : Windows 10 Pro 20H2

Unable to retrieve package list from http://www.nuget.org/api/v2/FindPackagesById()?id='System.Runtime.CompilerServices.Unsafe'&$orderby=Version asc
System.Net.WebException: The remote server returned an error: (400) Bad Request.
  at System.Net.HttpWebRequest+<GetResponseFromData>d__240.MoveNext () [0x00152] in <aa976c2104104b7ca9e1785715722c9d>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Net.HttpWebRequest.GetResponse () [0x00013] in <aa976c2104104b7ca9e1785715722c9d>:0 
  at NugetForUnity.NugetHelper.RequestUrl (System.String url, System.String userName, System.String password, System.Nullable`1[T] timeOut) [0x000b9] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.GetPackagesFromUrl (System.String url, System.String username, System.String password) [0x0006c] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.FindPackagesById (NugetForUnity.NugetPackageIdentifier package) [0x000ee] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
UnityEngine.Debug:LogErrorFormat (string,object[])
NugetForUnity.NugetPackageSource:FindPackagesById (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetPackageSource:GetSpecificPackage (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetHelper:GetOnlinePackage (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetHelper:GetSpecificPackage (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetWindow:DrawPackage (NugetForUnity.NugetPackage,UnityEngine.GUIStyle,UnityEngine.GUIStyle)
NugetForUnity.NugetWindow:DrawPackages (System.Collections.Generic.List`1<NugetForUnity.NugetPackage>)
NugetForUnity.NugetWindow:DrawOnline ()
NugetForUnity.NugetWindow:OnGUI ()
UnityEngine.GUIUtility:ProcessEvent (int,intptr,bool&)

**Comment by cgleadall** [comment]
One thing that I have noticed after clearing out my "Packages" folder...
Any NuGet package that has NO DEPENDENCIES will be downloaded to my "Packages" folder.

I tried to fool the 'packages.config' file and put all the packages in their dependency order thinking that if I explicitly get "Newtonsoft.Json" before I try to get the package-that-uses Newtonsoft I might avoid the searching of the dependencies... That did NOT work, but I noticed that only the packages with dependencies failed.

**Comment by opexsdc** [comment]
Please see the related PR.  

If you want to apply the fix and use this in Unity now, clone the repo, modify the request in ```Assets/NuGet/Editor/NugetPackageSource.cs``` to drop ```&$orderby=Version asc```.  Rebuild the using the repo ```build.ps1``` script to generate a new DLL.

(You may need to modify: ```Packager/ProjectSettings/ProjectVersion.txt``` to match whatever Unity version you have installed)

See: https://devblogs.microsoft.com/nuget/custom-v2-odata-queries-will-be-deprecated-march-9-2021/


----
There is probably some more sensible changes to be done to update to some newer version of OData - I'll leave that to wiser heads! :smile:

**Comment by jwittner** [comment]
Closed by #384 

Any users seeing this issue will likely see errors about comparing versions still, fixes for those are proposed in #385  - would really appreciate any testing against that PR by users seeing this issue. Thanks for all your support and contributions!

Note, not planning on making a new release until I can get fixes for the issues that #385 is addressing through as well. =)

**Comment by kalucky0** [comment]
I'm still getting the error
```
Unable to retrieve package list from http://www.nuget.org/api/v2/FindPackagesById()?id='ImpromptuNinjas.UltralightSharp.Core'&$orderby=Version asc
System.Net.WebException: The remote server returned an error: (400) Bad Request.
  at System.Net.HttpWebRequest+<GetResponseFromData>d__240.MoveNext () [0x00152] in <aa976c2104104b7ca9e1785715722c9d>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Net.HttpWebRequest.GetResponse () [0x00013] in <aa976c2104104b7ca9e1785715722c9d>:0 
  at NugetForUnity.NugetHelper.RequestUrl (System.String url, System.String userName, System.String password, System.Nullable`1[T] timeOut) [0x000b9] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.GetPackagesFromUrl (System.String url, System.String username, System.String password) [0x0006c] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
  at NugetForUnity.NugetPackageSource.FindPackagesById (NugetForUnity.NugetPackageIdentifier package) [0x000ee] in <ebb7710be32040ba9a0cb144ce2a8967>:0 
UnityEngine.Debug:LogErrorFormat (string,object[])
NugetForUnity.NugetPackageSource:FindPackagesById (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetPackageSource:GetSpecificPackage (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetHelper:GetOnlinePackage (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetHelper:GetSpecificPackage (NugetForUnity.NugetPackageIdentifier)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetHelper:Install (NugetForUnity.NugetPackage,bool)
NugetForUnity.NugetHelper:InstallIdentifier (NugetForUnity.NugetPackageIdentifier,bool)
NugetForUnity.NugetWindow:DrawPackage (NugetForUnity.NugetPackage,UnityEngine.GUIStyle,UnityEngine.GUIStyle)
NugetForUnity.NugetWindow:DrawPackages (System.Collections.Generic.List`1<NugetForUnity.NugetPackage>)
NugetForUnity.NugetWindow:DrawOnline ()
NugetForUnity.NugetWindow:OnGUI ()
UnityEngine.GUIUtility:ProcessEvent (int,intptr,bool&)
```

* **Nuget Package** : ImpromptuNinjas.UltralightSharp
* **NugetForUnity** : 3.0.1
* **Unity Version** : 2020.3.0f1
* **OS** : Windows 7 Ultimate

**Comment by jwittner** [comment]
@kalucky0  There wasn't a new release so if you're still testing from the releases you'll still see this. If you pull down the repo and update from that then it should be fixed.

**Comment by tobias-tengler** [comment]
Looking forward to the release of this fix. 
I would love to help testing and getting this released faster. Could you guide me to how I can install this package from source? :)

**Comment by DWishR** [comment]
Installing from source is easy, just drop [this folder](https://github.com/GlitchEnzo/NuGetForUnity/tree/master/Assets/NuGet) (and contents) into the Assets folder of your project. If you want #385 and #389 then use [this folder](https://github.com/DWishR/NuGetForUnity/tree/dev/directoryEntries/Assets/NuGet).

I want ahead and posted a Unity Package at [this comment](https://github.com/GlitchEnzo/NuGetForUnity/pull/389#issuecomment-820699071).

**Comment by jwittner** [comment]
Heads up that I've merged in #385 .

**Comment by shadiakiki1986** [comment]
Tested the posted unity package at dwishr's comment. Works fine on Unity 2021.02
